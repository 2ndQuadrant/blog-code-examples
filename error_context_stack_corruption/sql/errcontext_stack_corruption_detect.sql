\set client_min_messages notice

-- This is fine, normal return path
SELECT errcontext_stack_corruption.my_func(true, 'errcontext_detect');

-- This returns with a bogus errcontext, but if the compiler supports
-- the cleanup attribute it'll detect and clean up the problem instead
-- of corrupting the error context stack.
SELECT errcontext_stack_corruption.my_func(false, 'errcontext_detect');

-- Calling this should't crash if the compiler did the cleanup above
--
DO LANGUAGE plpgsql
$$
BEGIN
	RAISE EXCEPTION 'foo';
END;
$$;
