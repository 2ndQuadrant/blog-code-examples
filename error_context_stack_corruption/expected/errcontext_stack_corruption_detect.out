\set client_min_messages notice
-- This is fine, normal return path
SELECT errcontext_stack_corruption.my_func(true, 'errcontext_detect');
INFO:  did the thing
INFO:  after return from requested call
 my_func 
---------
 
(1 row)

-- This returns with a bogus errcontext, but if the compiler supports
-- the cleanup attribute it'll detect and clean up the problem instead
-- of corrupting the error context stack.
SELECT errcontext_stack_corruption.my_func(false, 'errcontext_detect');
WARNING:  not doing it and leaking the error context callback pointer!
ERROR:  leaked an errcontext callback pointer
-- Calling this should't crash if the compiler did the cleanup above
--
DO LANGUAGE plpgsql
$$
BEGIN
	RAISE EXCEPTION 'foo';
END;
$$;
ERROR:  foo
CONTEXT:  PL/pgSQL function inline_code_block line 3 at RAISE
